### ISSUE TITLE
Add numerically-stable function for angle between vectors

### ISSUE THREAD

**Issue by carlosgmartin** [pytorch/pytorch#59194]
## 🚀 Feature

Add a numerically-stable function that returns the angle between two vectors.

## Motivation

The most well-known formula for computing the angle between two vectors is
```python3
def angle_v0(a, b, dim=-1):
    c = (a * b).sum(dim) / (a.norm(dim=dim) * b.norm(dim=dim))
    return c.arccos()
```

Unfortunately, this formula is *extremely* numerically unstable when `a` and `b` are close to parallel or antiparallel. In fact, as `c` approaches ±1, the gradient of `c.arccos()` explodes to -∞. This is only an apparent pathology, however, since `∠(a, b)` itself is actually well-defined and well-behaved everywhere besides the origin.

## Alternatives

A clunky workaround is to clip the argument of `arccos`, such as
```python3
def angle_v1(a, b, dim=-1, eps=1e-4):
    z = (a * b).sum(dim) / (a.norm(dim=dim) * b.norm(dim=dim))
    return z.clip(-1 + eps, 1 - eps).arccos()
```
However:
1. If `eps` is large, this compromises the correctness of the result.
2. If `eps` is small, this is still numerically unstable.

## Pitch

See [William Kahan](https://en.wikipedia.org/wiki/William_Kahan)'s paper [*How Futile are Mindless Assessments of Roundoff in Floating-Point Computation? §12: Mangled Angles*](https://people.eecs.berkeley.edu/~wkahan/Mindless.pdf#page=46):

> We shall consider now a simpler example chosen to support the thesis that numerical software is extremely difficult to debug.  Evidence for this thesis is the longevity of inaccurate software in use by vast numbers of numerically active and, in most instances, sophisticated users of MATLAB. Once again we consider the angle between two directions x and y specified now in Cartesian coordinates
> ...
> Here is a better formula less well known than it deserves: ∠(x, y) := 2 arctan( ||x·||y|| – ||x||·y||/||x·||y|| + ||x||·y|| ). When executed in arithmetic rounded to n significant bits its absolute error never much exceeds 1/2ⁿ⁻¹ unless the dimension is gargantuan. And redirected roundoff barely affects this formula.
> ...
> What do mangled angles teach us? The goal of error-analysis is not to find errors but to fix them.  They have to be found first. The embarrassing longevity, over three decades, of inaccurate and/or ugly programs to compute a function so widely used as ∠(X, Y) says something bleak about the difficulty of floating-point error-analysis for experts and nonexperts: Without adequate aids like redirected roundings, diagnosis and cure are becoming practically impossible. 

Here is this formula implemented in PyTorch:

```python3
def angle_v2(a, b, dim=-1):
    x_norm = a.norm(dim=dim, keepdim=True)
    y_norm = a.norm(dim=dim, keepdim=True)
    return 2 * torch.atan2(
        (a * y_norm - x_norm * b).norm(dim=dim),
        (a * y_norm + x_norm * b).norm(dim=dim)
    )
```

The proposal is to add this to the PyTorch standard library so that users can benefit from access to a numerically-stable version of this fundamental geometric function, without having to deal with the issues that arise from the naive formula.

## Additional context

See also the Math StackExchange question [*Numerically stable method for angle between 3D vectors*](https://math.stackexchange.com/questions/1143354/numerically-stable-method-for-angle-between-3d-vectors).

**Comment by engrmz** [comment]
Hi,

In version 2, both x_norm and y_norm are same? 

`    x_norm = a.norm(dim=dim, keepdim=True)`
`    y_norm = a.norm(dim=dim, keepdim=True)`

-mz

**Comment by carlosgmartin** [comment]
@engrmz Fixed.

**Comment by IanMaquignaz** [comment]
Stumbled across this and though I should write-up my findings in case anyone else was in my situation. 

I create the following script to test different methods, focusing on a desired dtype for `float32` with vectors on a unit circle: 
```python

ok_count = 0
overall_min = torch.tensor([float('inf')], dtype=torch.float64)
overall_max = torch.tensor([float('-inf')], dtype=torch.float64)
for angle in torch.linspace(
    -4*torch.pi, 4*torch.pi,
    steps=1000,
    dtype=torch.float32,
    device=None,
    requires_grad=False,
):

    x = torch.cos(angle)
    y = torch.sin(angle)

    def torch_atan2_angle(x, y):
        return torch.atan2(y, x)

    def torch_complex_angle(x, y):
        return torch.angle(torch.complex(x, y))

    def mindless_1(x, y, dim=-1):
        # x' is x inverse
        # ∠(x, y) := if x'·y ≥ 0 then arcsin( ||xxy||/(||x||·||y||) ) else π – arcsin( ||xxy||/(||x||·||y|
        x_ = x*-1
        x_norm = x.norm(dim=dim, keepdim=True)
        y_norm = y.norm(dim=dim, keepdim=True)
        a =(torch.dot(x_, y)/torch.dot(x_norm, y_norm)).clamp(min=-1, max=1)
        return torch.arccos(a)

    def mindless_2(x, y, dim=-1):
        # cross-product angle
        # (x, y) := if x'·y ≥ 0 then arcsin( ||xxy||/(||x||·||y||) ) else π – arcsin( ||xxy||/(||x||·||y||)
        cp = torch.linalg.cross(x, y)
        norm_x = torch.linalg.norm(x)
        norm_y = torch.linalg.norm(y)
        norm_cp = torch.linalg.norm(cp)
        sin_theta = (norm_cp / (norm_x * norm_y))
        angle = torch.arcsin(sin_theta)
        if torch.dot(x*-1, y) >= 0:
            return angle
        else:
            return torch.pi - angle

    def mindless_3(x, y, dim=-1):
        x_norm = x.norm(dim=dim, keepdim=True)
        y_norm = y.norm(dim=dim, keepdim=True)
        return 2. * torch.arctan(
            (x * y_norm - x_norm * y).norm(dim=dim)/(x * y_norm + x_norm * y).norm(dim=dim)
        )

    def angle_v2(a, b, dim=-1):
        a_norm = a.norm(dim=dim, keepdim=True)
        b_norm = b.norm(dim=dim, keepdim=True)
        return 2 * torch.atan2(
            (a * b_norm - a_norm * b).norm(dim=dim),
            (a * b_norm + a_norm * b).norm(dim=dim)
        )

    vector = torch.stack([x,y, torch.tensor(0., dtype=x.dtype)], dim=0)
    vector_ref = torch.tensor([1., 0., 0], dtype=x.dtype)

    # Pick a Method to Reconstruct Angle
    angle_recon = torch_atan2_angle(x,y)
    angle_recon = torch_complex_angle(x,y)
    angle_recon = mindless_1(vector_ref, vector)
    angle_recon = mindless_2(vector_ref, vector)
    angle_recon = mindless_3(vector_ref, vector)
    angle_recon = angle_v2(vector_ref, vector)

    angle_real = torch.fmod(angle, 2*torch.pi)
    angle_recon = torch.fmod(angle_recon, 2*torch.pi)
    diff = [
        torch.abs(angle_real - angle_recon),
        torch.abs((angle_real - 2*torch.pi) - angle_recon),
        torch.abs((angle_real + 2*torch.pi) - angle_recon),
    ]
    min_diff = min(diff)
    overall_min = torch.minimum(overall_min, min_diff)
    overall_max = torch.maximum(overall_max, min_diff)
    if min_diff == overall_max:
        print(f"angle: {angle_real}, a: {angle_recon}, diff_min: {min_diff}")
    else:
        ok_count+=1
    # assert torch.allclose(min(diff), torch.tensor(0., dtype=x.dtype), atol=1e1), f"angle: {angle_real}, a: {angle_recon}, diff: {diff}"
    # print('----')
print(f"Min error: {overall_min.item()}, Max error: {overall_max.item()}")
print(f"Number of successful cases: {ok_count} / 1000")
```

What this found:
`torch_atan2_angle`  --> Min error: 0.0, Max error: 4.76837158203125e-07
`torch_complex_angle`   --> Min error: 0.0, Max error: 4.76837158203125e-07
[ angle_v2 method proposed by @carlosgmartin as `angle_v2` ](https://github.com/carlosgmartin)]{https://github.com/pytorch/pytorch/issues/59194#issue-907097096) --> Min error: 0.0, Max error: 3.13844895362854
[ Mindless 1 from source doc  pg.46-47 ](https://people.eecs.berkeley.edu/~wkahan/Mindless.pdf)  --> Min error: 0.0031441450119018555, Max error: 3.1415927410125732
[ Mindless 2 from source doc  pg.46-47 ](https://people.eecs.berkeley.edu/~wkahan/Mindless.pdf)  --> Min error: 0.003116130828857422, Max error: 3.1415927410125732
[ Mindless 3 from source doc pg.46-47 ](https://people.eecs.berkeley.edu/~wkahan/Mindless.pdf)  --> Min error: 0.0, Max error: 3.13844895362854

Therefore, assuming no errors in my implementation, the proposed method and the other methods suggested by the source suffer significant errors with nearly (anti-)parallel vectors. This matches the discussion in the source, which states `These formulas lose about half the digits carried when ∠(x, y) ≈ π/2 '.

In conclusion, the existing pytorch implementations (tested as `torch_atan2_angle` and `torch_complex_angle`) offer reliable accuracy.
